### PR TITLE
Delete runs from DR servers

### PR DESCRIPTION
**Pull request by jenniferliddle**

### CODE REVIEW

File: bin/npg_delete_dr_runs [hunk 1]
```diff
@@ -14,51 +14,98 @@ use npg_tracking::Schema;
...
 while (my $run=$rs->next()) {
     my $status = $run->current_run_status_description || "";
     my $tag = $run->is_tag_set("staging");
     my $run_folder = $run->folder_name || "";
     if ($run_folder and !$tag ) {
-        print $run_folder, "\t", $run->id_run, "\t", $run->folder_path_glob, "\t", $status, "\n";
+        my $prefix = substr($run_folder,0,4);
+        my $dname = "/exports/$prefix/$run_folder";
+        if ($files and !$folders) {
+            $cmd .= "if [ -d $dname ]\\\; then rm $dname/Data/Intensities/BaseCalls/L\*/C\*/\*.cbcl\\\; fi\\\;";
```
Comment (dkj) [inline bin/npg_delete_dr_runs]: I think we should be able to drop the second level of escaping

File: bin/npg_delete_dr_runs [hunk 2]
```diff
@@ -14,51 +14,98 @@ use npg_tracking::Schema;
...
+            $cmd .= "if [ -d $dname ]\\\; then rm -r $dname\\\; fi\\\;";
+        }
+        if (!$files and !$folders) {
+            $cmd .= "if [ -d $dname ]\\\; then ls -ld $dname\\\; fi\\\;";
+        }
     }
 }
+chop $cmd;
+if ($no_dry_run) {
+    system("echo $cmd | ssh ukbb\@seq-stor-01 bash");
```
Comment (dkj) [inline bin/npg_delete_dr_runs]: We could open filehandles for each `ssh` process forked and write commands run folder by runfolder to each filehandle. Printing to those avoids the `echo` and extra level of escaping. 
I fear `echo $cmd` might grow too large for the shell.
What about running `bash -e` as we're not expecting failures given the testing for directories.

Comment (mgcam) [reply bin/npg_delete_dr_runs]: There is more to closing filehandles like this than is currently in the code, see https://perldoc.perl.org/functions/close.html

File: bin/npg_delete_dr_runs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+our $VERSION = '0';
+
+my $help;
+my $files;
+my $folders;
+my $no_dry_run;
+
+sub send_commands {
+    my ($dr_server, $cmd) = @_;
+    open(BASH, "| ssh $dr_server bash -e") or die("Can't run ssh: $!");
```
Comment (dkj) [inline bin/npg_delete_dr_runs]: Would prefer the `open FILEHANDLE,MODE,EXPR,LIST` form
We use `croak` elsewhere in this script.
```suggestion
    open(BASH, '|-', 'ssh', $dr_server, qw(bash -e)) or croak("Can't run ssh: $!");
```
(untested)

File: bin/npg_delete_dr_runs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+pod2usage(0) if $help;
+
+my $rs=npg_tracking::Schema->connect()->resultset(q(RunStatusDict))
+    ->search({description=>[q[run cancelled],q[data discarded],q[qc complete]]})
+    ->related_resultset(q(run_statuses))->search({q(run_statuses.iscurrent)=>1})
+    ->related_resultset("run")->search({q(run.folder_path_glob)=>{LIKE=>q(%/esa-%)}});
+
+my $cmd = '';
+while (my $run=$rs->next()) {
+    my $status = $run->current_run_status_description || "";
```
Comment (dkj) [inline bin/npg_delete_dr_runs]: unused
```suggestion
```

File: bin/npg_delete_dr_runs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            $cmd .= "if [ -d $dname ]; then rm -r $dname; fi\n";
+        }
+        if (!$files and !$folders) {
+            $cmd .= "if [ -d $dname ]; then ls -ld $dname; fi\n";
+        }
+    }
+}
+chop $cmd;
+if ($no_dry_run) {
+    send_commands('ukbb\@seq-stor-01', $cmd);
```
Comment (dkj) [inline bin/npg_delete_dr_runs]: Uncomfortable with hostnames hardcoded (in this repo) - if staying in `npg_tracking` these should be able to be set from command line options (but they could be defaults - but include as fqdn if so).

**Comment by dkj** [comment]
We could probably drop the cbcl only option now.

**Comment by dkj** [comment]
We should reconsider integrating the hard linking for data products and cached top-up candidates.

File: bin/npg_delete_dr_runs (hunk 5 above)
Comment (jenniferliddle) [reply bin/npg_delete_dr_runs]: Simplifying things is good. I like simple.

**Comment by jenniferliddle** [comment]
> We could probably drop the cbcl only option now.

We _could_,  but I'd be inclined to keep the option in for now at least. There is nothing to say we have to use it.

**Comment by jenniferliddle** [comment]
> We should reconsider integrating the hard linking for data products and cached top-up candidates.

Sounds good. Do we have code for hard linking for data products and cached top-up candidates to integrate? Where is it?

**Comment by jenniferliddle** [comment]
Being re-written and moving to another repository.

**Pull request closed without merging**